src/lib/mdns/Discovery_ImplPlatform.cpp

```diff
@@ -154,100 +121,102 @@ CHIP_ERROR DiscoveryManager::SetupHostname()
-    snprintf(service.mName, sizeof(service.mName), "%016" PRIX64, mUnprovisionedInstanceName);
+    char discriminatorBuf[6];
+    char vendorProductBuf[12];
+    TextEntry textEntries[3];
+    size_t textEntrySize = 0;
+    char shortDiscriminatorSubtype[6];
+    char longDiscriminatorSubtype[7];
+    char vendorSubType[8];
+    const char * subTypes[3];
+    size_t subTypeSize = 0;
+
+    if (!mMdnsInitialized)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    snprintf(service.mName, sizeof(service.mName), "%016" PRIX64, mCommissionInstanceName);
     strncpy(service.mType, "_chipc", sizeof(service.mType));
     service.mProtocol = MdnsServiceProtocol::kMdnsProtocolUdp;
-    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetVendorId(vendorID));
-    SuccessOrExit(error = chip::DeviceLayer::ConfigurationMgr().GetProductId(productID));
-    snprintf(discriminatorBuf, sizeof(discriminatorBuf), "%04X", discriminator);
-    snprintf(vendorProductBuf, sizeof(vendorProductBuf), "%04X+%04X", vendorID, productID);
-    entries[0].mData     = reinterpret_cast<const uint8_t *>(discriminatorBuf);
-    entries[0].mDataSize = strnlen(discriminatorBuf, sizeof(discriminatorBuf));
-    entries[1].mData     = reinterpret_cast<const uint8_t *>(vendorProductBuf);
-    entries[1].mDataSize = strnlen(discriminatorBuf, sizeof(vendorProductBuf));
 
+    snprintf(discriminatorBuf, sizeof(discriminatorBuf), "%04u", params.GetLongDiscriminator());
+    textEntries[textEntrySize++] = { "D", reinterpret_cast<const uint8_t *>(discriminatorBuf),
+                                     strnlen(discriminatorBuf, sizeof(discriminatorBuf)) };
+    if (params.GetVendorId().HasValue())
+    {
+        if (params.GetProductId().HasValue())
+        {
+            snprintf(vendorProductBuf, sizeof(vendorProductBuf), "%u+%u", params.GetVendorId().Value(),
+                     params.GetProductId().Value());
+        }
+        else
+        {
+            snprintf(vendorProductBuf, sizeof(vendorProductBuf), "%u", params.GetVendorId().Value());
+        }
+        textEntries[textEntrySize++] = { "VP", reinterpret_cast<const uint8_t *>(vendorProductBuf),
+                                         strnlen(vendorProductBuf, sizeof(vendorProductBuf)) };
+    }
 #if CHIP_ENABLE_ADDITIONAL_DATA_ADVERTISING
-    // Rotating Device ID
-    entries[2].mData     = reinterpret_cast<const uint8_t *>(CHIP_ROTATING_DEVICE_ID);
-    entries[2].mDataSize = strnlen(CHIP_ROTATING_DEVICE_ID, sizeof(CHIP_ROTATING_DEVICE_ID));
+    textEntries[textEntrySize++] = { "RI", reinterpret_cast<const uint8_t *>(CHIP_ROTATING_DEVICE_ID),
+                                     strlen(CHIP_ROTATING_DEVICE_ID) };
 #endif
-    service.mTextEntryies  = entries;
-    service.mTextEntrySize = sizeof(entries) / sizeof(TextEntry);
+
+    snprintf(shortDiscriminatorSubtype, sizeof(shortDiscriminatorSubtype), "_S%03u", params.GetShortDiscriminator());
```

Comment:
Avahi requires all the subtypes to start with an underscore. Since this is also the convention for all dns services, shall we update the spec to use `_Sxxx` rather than `Sxxx` for subtype names?

Reply:
Chip spec does not use _. Maybe question for Stewart? the example mac-tools did not use underscore either.

Reply:
https://tools.ietf.org/html/rfc6763 says:

`Subtype strings are not required to begin with an underscore, though they often do.`

Reply:
@gjc13 - could you create a GH Issue for this and we can ask Stewart to comment?

We would list things like 'avahi does not allow sub names without underscore' and we need a decision on what to do (use _ in the spec, never use avahi (ouch), try to patch avahi, something else).

Reply:
Created https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/852 and assigned to Matt and Stuart for comment. Hoping that using underscores is reasonable.